test-utils/src/main/java/io/littlehorse/test/LHExtension.java

```diff
@@ -70,4 +82,46 @@ public void postProcessTestInstance(Object testInstance, ExtensionContext contex
         }
         testContext.instrument(testInstance);
     }
+
+    private void maybeCreateTenantAndPrincipal(TestContext testContext) {
+        Optional<String> principalId = getPrincipalId();
+        if (principalId.isPresent()) {
+            try {
+                testContext
+                        .getAnonymousClient()
+                        .putTenant(PutTenantRequest.newBuilder()
+                                .setId(testContext.getConfig().getTenantId())
+                                .build());
+                ServerACLs acls = ServerACLs.newBuilder()
+                        .addAcls(ServerACL.newBuilder()
+                                .addAllowedActions(ACLAction.ALL_ACTIONS)
+                                .addResources(ACLResource.ACL_ALL_RESOURCES)
+                                .build())
+                        .build();
+                testContext
+                        .getAnonymousClient()
+                        .putPrincipal(PutPrincipalRequest.newBuilder()
+                                .setId(principalId.get())
+                                .setGlobalAcls(acls)
+                                .putPerTenantAcls(testContext.getConfig().getTenantId(), acls)
+                                .build());
+                // wait until the server is up
+                Awaitility.await()
+                        .atMost(Duration.ofSeconds(15))
+                        .ignoreException(RuntimeException.class)
+                        .until(() -> {
+                            Principal whoami = testContext.getLhClient().whoami(Empty.getDefaultInstance());
+                            return whoami.getId().getId().equals(principalId.get());
+                        });
+            } catch (StatusRuntimeException ex) {
+                if (!ex.getStatus().getCode().equals(Status.Code.ALREADY_EXISTS)) {
+                    throw ex;
+                }
+            }
+        }
+    }
+
+    private Optional<String> getPrincipalId() {
+        return Optional.ofNullable(System.getenv().getOrDefault("LH_CLIENT_ID", "my-principal"));
```

Comment:
There is no point on returning a Optional here. I'll change this

Reply:
Why not `Optional<Optional<Optional<String>>>`?

But anyways, I would prefer to see it return `PrincipalIdModel` though.